File: wk4/python/trappingrainwater.py
#link: https://leetcode.com/problems/trapping-rain-water
class Solution:
    def trap(self, height: List[int]) -> int:  
        total = 0

        left, right = 0, len(height) - 1
        leftMax = height[left]
        rightMax = height[right]

        while left < right:
            if leftMax <= rightMax:
                total += leftMax - height[left]
                left += 1
                leftMax = max(leftMax, height[left])
            else:
                total += rightMax - height[right]
                right -= 1
                rightMax = max(rightMax, height[right])
        return total

File: wk4/java/trappingRainWater.java
// problem link: https://leetcode.com/problems/trapping-rain-water/description/

class Solution {
    public int trap(int[] height) {
        int l = 0, r = height.length-1, lmax = height[l], rmax = height[r], ret = 0;
        while (l < r) {
            if(lmax < rmax) {
                l++;
                lmax = Math.max(lmax, height[l]);
                ret += lmax - height[l];
            }
            else {
                r--;
                rmax = Math.max(rmax, height[r]);
                ret += rmax - height[r];
            }
        }
        return ret;
    }
}